### PR TITLE
BRS for comosDB minimum required TLS version

### PR DESCRIPTION
**Pull request by MJaithra**

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 1]
```diff
@@ -0,0 +1,975 @@
...
+
+            Write-Host "[Step 4 of 4] Configure required minimum TLS version on the CosmosDB Accounts in the Subscription..."
+            Write-Host $([Constants]::SingleDashLine)
+
+            if (-not $Force)
+            {
+                Write-Host "Do you want to set minimum TLS version to required minimum TLS version for all CosmosDB Account(s)? " -ForegroundColor $([Constants]::MessageType.Warning) -NoNewline
+        
+
+                $userInput = Read-Host -Prompt "(Y|N)" #TODO:
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:525]: nothing, that was by mistake. I'll correct it.

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 2]
```diff
@@ -0,0 +1,975 @@
...
+
+    if($AutoRemediation)
+    {
+        if(-not (Test-Path -Path $Path))
+        {
+        Write-Host "File containing failing controls details [$($Path)] not found. Skipping remediation..." -ForegroundColor $([Constants]::MessageType.Error)
+        Write-Host $([Constants]::DoubleDashLine)
+        return
+        }
+        Write-Host "Fetching all CosmosDB Accounts failing for the [$($controlIds)] control from [$($Path)]..." -ForegroundColor $([Constants]::MessageType.Info)
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:286]: sure, done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 3]
```diff
@@ -0,0 +1,975 @@
...
+    if(-not($AutoRemediation))	
+    {
+        Write-Host "Subscription Name: [$($context.Subscription.Name)]"
+        Write-Host "Subscription ID: [$($context.Subscription.SubscriptionId)]"
+        Write-Host "Account Name: [$($context.Account.Id)]"
+        Write-Host "Account Type: [$($context.Account.Type)]"
+        Write-Host $([Constants]::SingleDashLine)
+    }
+
+    Write-Host "To Set Minimum TLS version for CosmosDB Accounts in a Subscription, Contributor or higher privileges on the CosmosDB Accounts are required." -ForegroundColor $([Constants]::MessageType.Warning)
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:261]: Done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 4]
```diff
@@ -0,0 +1,975 @@
...
+    Use approved version of TLS for Azure CosmosDB.
+
+# Prerequisites:
+    1. Contributor or higher privileges on the CosmosDB Accounts in a Subscription.
+    2. Must be connected to Azure with an authenticated account.
+
+# Steps performed by the script:
+    To remediate:
+        1. Validate and install the modules required to run the script and validate the user.
+        2. Get the list of CosmosDB Accounts in a Subscription that do not use the required minimum TLS version.
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:18]: sure, done

File: Scripts/RemediationScripts/Readme.md [hunk 5]
```diff
@@ -1016,6 +1017,23 @@ Yes
 
 ___ 
 
+## 49. 	Azure_CosmosDB_DP_Use_Secure_TLS_Version
+
+### Display Name
+Use approved version of TLS for the CosmosDB
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Readme.md:1023]: done

File: Control coverage/Feature/CosmosDB.md [hunk 6]
```diff
@@ -269,3 +270,48 @@ Threat Protection for Azure Cosmos DB provides an additional layer of security i
...
+> 
+> **Failed:** 
+> MinTLSVersion is lessthan 1.2 for CosmosDB.
+> 
+
+### Recommendation 
+
+- **Azure Portal** 
+
+	 To set required TLS version: Go to Azure Portal --> your CosmosDB --> Settings --> Networking --> Connectivity --> Minimum Transport layer security protocol dropdown. Select the required TLS version from the dropdown.
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Control coverage/Feature/CosmosDB.md:293]: ok, done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 7]
```diff
@@ -0,0 +1,975 @@
...
+    This script is used to set required minimum TLS version for CosmosDB Accounts in a Subscription.
+
+# Control ID:
+    Azure_CosmosDB_DP_Use_Secure_TLS_Version
+
+# Display Name:
+    Use approved version of TLS for Azure CosmosDB.
+
+# Prerequisites:
+    1. Contributor or higher privileges on the CosmosDB Accounts in a Subscription.
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:12]: done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 8]
```diff
@@ -0,0 +1,975 @@
...
+        .PARAMETER SkipBackup
+        Specifies that no back up will be taken by the script before remediation.
+        
+        .PARAMETER FilePath
+        Specifies the path to the file to be used as input for the remediation.
+
+        .PARAMETER Path
+        Specifies the path to the file to be used as input for the remediation when AutoRemediation switch is used.
+
+        .PARAMETER AutoRemediation
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:146]: Added required commands in main file.

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 9]
```diff
@@ -0,0 +1,975 @@
...
+        Specifies script is run as a subroutine of AutoRemediation Script.
+
+        .PARAMETER TimeStamp
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:149]: Added required code in the main file

File: Control coverage/Feature/CosmosDB.md [hunk 10]
```diff
@@ -269,3 +270,48 @@ Threat Protection for Azure Cosmos DB provides an additional layer of security i
 
 ___ 
 
+## Azure_CosmosDB_DP_Use_Secure_TLS_Version
+
+### Display Name 
+Use approved version of TLS for the CosmosDB
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Control coverage/Feature/CosmosDB.md:276]: Done

File: Control coverage/Feature/CosmosDB.md [hunk 11]
```diff
@@ -269,3 +270,48 @@ Threat Protection for Azure Cosmos DB provides an additional layer of security i
...
+
+### Rationale 
+TLS provides confidentiality and data integrity between client and server. Using approved TLS version significantly reduces risks from security design issues and security bugs that may be present in older versions.
+### Control Spec 
+
+> **Passed:** 
+> MinTLSVersion is 1.2 for CosmosDB.
```
[the remark this answers is not in the document]
Comment (MJaithra) [reply Control coverage/Feature/CosmosDB.md]: Added control settings and verified other docs for reference on control spec.

File: Control coverage/Feature/CosmosDB.md [hunk 12]
```diff
@@ -269,3 +270,56 @@ Threat Protection for Azure Cosmos DB provides an additional layer of security i
...
+### Control Settings 
+```json 
+{
+    "MinReqTLSVersion": "1.2"
+}
+ ```  
+### Control Spec 
+
+> **Passed:** 
+> Minimum TLS version is set to 1.2 or higher for Cosmos DB.
```
Comment (jaiswalricha) [inline Control coverage/Feature/CosmosDB.md:290]: Minimum TLS version is set to required or higher for Cosmos DB.

File: Control coverage/Feature/CosmosDB.md [hunk 13]
```diff
@@ -269,3 +270,56 @@ Threat Protection for Azure Cosmos DB provides an additional layer of security i
...
+> 
+> **Failed:** 
+> Any of the following conditions is met.
+> * Minimum TLS version is not set (default 1.0).
+> * Minimum TLS version is less than 1.2 (configured min required TLS version) for Cosmos DB.
```
Comment (jaiswalricha) [inline Control coverage/Feature/CosmosDB.md:295]: same as above. let's not mention 1.2 but req only

File: Scripts/RemediationScripts/Readme.md (hunk 5 above)
Comment (jaiswalricha) [inline Scripts/RemediationScripts/Readme.md:1023]: space missing b/w cosmos and db

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 14]
```diff
@@ -0,0 +1,975 @@
...
+    Write-Host "[Step 3 of 4] Back up Cosmos DB Account(s) details..."
+    Write-Host $([Constants]::SingleDashLine)
+    if ([String]::IsNullOrWhiteSpace($FilePath))
+    {        
+        if(-not $SkipBackup)
+        {
+            # Backing up Cosmos DB Account details.
+            $backupFile = "$($backupFolderPath)\CosmosDBAccountsWithoutReqMinTLSVersion.csv"
+            $CosmosDBAccountsWithoutReqMinTLSVersion | Export-CSV -Path $backupFile -NoTypeInformation
+            Write-Host "Cosmos DB Account(s) details have been successful backed up to [$($backupFile)]" -ForegroundColor $([Constants]::MessageType.Update)
```
Comment (jaiswalricha) [inline Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:500]: successfully**

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 [hunk 15]
```diff
@@ -0,0 +1,975 @@
...
+        else
+        {
+            Write-Host $([Constants]::SingleDashLine)
+            Write-Host "Remediation Summary:" -ForegroundColor $([Constants]::MessageType.Info)
+            Write-Host $([Constants]::SingleDashLine)
+        
+            if ($($CosmosDBAccountsRemediated | Measure-Object).Count -gt 0)
+            {
+                Write-Host "Successfully set the Minimum TLS version to required Minimum TLS version on the following Cosmos DB Account(s) in the subscription:" -ForegroundColor $([Constants]::MessageType.Update)
+                Write-Host "Please note that it might take a while to update and reflect in the portal." -ForegroundColor $([Constants]::MessageType.Warning)
```
Comment (jaiswalricha) [inline Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:644]: reflect in the Azure Portal

File: Control coverage/Feature/CosmosDB.md (hunk 12 above)
Comment (MJaithra) [reply Control coverage/Feature/CosmosDB.md:290]: done

File: Control coverage/Feature/CosmosDB.md (hunk 13 above)
Comment (MJaithra) [reply Control coverage/Feature/CosmosDB.md:295]: done

File: Scripts/RemediationScripts/Readme.md (hunk 5 above)
Comment (MJaithra) [reply Scripts/RemediationScripts/Readme.md:1023]: done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 (hunk 14 above)
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1:500]: done

File: Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1 (hunk 15 above)
Comment (MJaithra) [reply Scripts/RemediationScripts/Remediate-SetCosmosDBAccountMinReqTLSVersion.ps1]: done

**Review by jaiswalricha — APPROVED** [review]

**Review by vaishnavipulluri — APPROVED** [review]

**Pull request merged**